### PR TITLE
busybox: disable PREFER_IPV4_ADDRESS

### PR DESCRIPTION
**Pull request by traud**
which breaks ipv6-only (bugs [#12167](https://dev.archive.openwrt.org/ticket/12167.html), [FS#84](https://bugs.openwrt.org/?do=details&task_id=84), and [FS#2608](https://bugs.openwrt.org/?do=details&task_id=2608)) and is [considered harmful](https://bugs.busybox.net/show_bug.cgi?id=12381). The original discussion of that bug report was about BusyBox `wget` (HTTP client) and driven within [Alpine Linux …](https://gitlab.alpinelinux.org/alpine/aports/-/issues/10937) and there was an [upstream discussion](http://lists.busybox.net/pipermail/busybox/2016-August/084570.html) about BusyBox `ntpd` (NTP client). However, in that bug report, there is one sentence:
> Unless the whole [BusyBox] networking framework [tries] fallbacks like `getaddrinfo` returns […]

The referenced [RFC 6724](https://tools.ietf.org/html/rfc6724) talks about lists and has a section 10.3.1 [Handling Broken IPv6](https://tools.ietf.org/html/rfc6724#section-10.3.1):
> [W]hen a host has IPv4 connectivity […] but has "broken" IPv6 connectivity […], this can result in unwanted timeouts.

By the way, the same is true for the opposite. Back to that list/fallback:
In BusyBox, its current `str2sockaddr(.)` and `host2sockaddr(.)` do not return a list of IP addresses but just a single IP address. Therefore, all (?) networking tools within BusyBox (except its `nslookup`) are not able to fallback to a next IP address, let alone [fast fallback](https://en.wikipedia.org/wiki/Happy_Eyeballs). Although this Pull Request here solves the IPv6-only scenario, it further defers a general fallback (and fast fallback) approach, which the BusyBox networking framework requires to be on par with modern operating systems.

### CODE REVIEW

**Comment by aparcar** [comment]
Looks good to me. 

> which the BusyBox networking framework requires to be on par with modern operating systems.

Are you aware of any such work?

**Comment by zorun** [comment]
I had a look and I think we should merge this, but it's a change in behaviour for busybox network tools (basically ping, traceroute, ntpd) : on a dual-stack router and for a name with A+AAAA, these tools currently use IPv4, while they would now use IPv6 with this patch. Not a big deal (and a good change in my opinion) but this needs to be noted.

TL;DR `getaddrinfo` in musl does the right thing since 2014 (musl 1.1.3), so the `PREFER_IPV4_ADDRESS` hack in busybox is no longer needed for IPv4-only hosts and is indeed actively breaking things for IPv6-only hosts.

I agree that Busybox has a non-standard approach because it should try all addresses returned by `getaddrinfo`, not just the first one (according to RFC6724). But fixing this looks like a big change in Busybox.

In any case, if the libc is well-behaved and `getaddrinfo` returns addresses in a sensible order, Busybox should be fine in most cases by just using the first returned address (which could be either IPv4 or IPv6).

Musl implements RFC 3484/6724 for `getaddrinfo` since 2014 (musl 1.1.3, commit 3c42605a603), so it is well-behaved. It sorts IPv6 addresses before IPv4 addresses by default, but it also checks reachability of each address (among other things). So, for an IPv4-only router that resolves a name with A+AAAA, musl should return IPv4 addresses first (this needs to be tested). This means that busybox will use the IPv4 address without needing the `PREFER_IPV4_ADDRESS` hack.

I will test this to make sure it works as expected, and then merge it in time for 21.02.

**Comment by aparcar** [comment]
> I will test this to make sure it works as expected, and then merge it in time for 21.02.

@zorun thanks for the research and for taking this over!

**Comment by zorun** [comment]
Merged in master with a more detailed commit message: https://git.openwrt.org/7fea9d9f5dd282a7049d77cc6b75e0a703ead26c

Thanks for fixing a 9-years-old bug in OpenWrt ;)

I will backport it to 21.02 in a few days if there is no fallout.

**Comment by traud** [comment]
> fixing a 9-years-old bug in OpenWrt

One, the first (?) bug report was from Sep. 2012, yes. However, this issue was introduced Nov. 2007 [in BusyBox](https://git.busybox.net/busybox/commit/networking/Config.in?id=9d6c46955fb51d7867e0d8d0ed94006ab72e4821) and Jun. 2008 [in OpenWrt](https://dev.archive.openwrt.org/attachment/ticket/3353/busybox-1.10.2-update2.patch). And connectivity is not a third-tier topic in OpenWrt, isn’t it? I wonder why it takes so long to detect such issues (bug report) and then to fix those issues. Furthermore, when the new bug tracker was created, that old issue was not transferred. Personally, I am using my OpenWrt as travel ‘router’. And in hotels, IPv6-only connectivity is quite common because their DHCPv4 server broke down but IPv6 continues to work.

>> which the BusyBox networking framework requires to be on par with modern operating systems.

> Are you aware of any such work?

You mean, whether there are other open-source projects having introduced (fast) fallback so their specification/behavior can be copied over avoiding duplicate development research? No, not aware of, but I have not even searched for.

Or are you about BusyBox itself and wether there is an issue report or a feature branch or a patch adding (fast) fallback to BusyBox networking tools? No, not aware of, but I have not even searched for.

That raises my next question, this time not rhetoric as my question above:
We noticed that (fast) fallback network connectivity is missing BusyBox. Who is going to (at least) report that upstream? Actually, I came upstream already because I am not on vanilla OpenWrt but on two different commercial distributions of OpenWrt – both do not care about IPv6 at all actually, yet, much, well, name it like you want. Anyway, I had to go upstream because there was no help. Do I have to go one step further upstream? Looks like the Alpine Linux folks (and their downstream projects) did not go either. Go figure!

**Comment by zorun** [comment]
It takes so long because openwrt/musl/busybox are community projects, no need to rant about it.

Note that musl's `getaddrinfo()` was only fixed in 2014. Before that, it could return AAAA records even on IPv4-only hosts, which likely explains why this hack was introduced in 2007/2008 (busybox needs to support broken libc). But the root issue is that busybox should try all addresses returned by `getaddrinfo()`, not just the first one. It looks like a significant change, so this definitely has to be proposed upstream by someone who cares.

**Comment by aparcar** [comment]
Missing PKG_RELEASE bump btw
